

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Analyze 2018 Pilot runs in Parallel &mdash; HTS2018 1.0 documentation</title>
  

  
  
  
  

  

  
  
    

  

  
    <link rel="stylesheet" href="../../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" /> 

  
  <script src="../../_static/js/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search">
          

          
            <a href="../../index.html" class="icon icon-home"> HTS2018
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">Slides:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../slides.html">Slide sets</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../index.html">HTS2018</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../../index.html">Docs</a> &raquo;</li>
        
      <li>Analyze 2018 Pilot runs in Parallel</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="../../_sources/josh/setup_and_testing/test_sort_RAM.ipynb.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  
<style>
/* CSS for nbsphinx extension */

/* remove conflicting styling from Sphinx themes */
div.nbinput,
div.nbinput div.prompt,
div.nbinput div.input_area,
div.nbinput div[class*=highlight],
div.nbinput div[class*=highlight] pre,
div.nboutput,
div.nbinput div.prompt,
div.nbinput div.output_area,
div.nboutput div[class*=highlight],
div.nboutput div[class*=highlight] pre {
    background: none;
    border: none;
    padding: 0 0;
    margin: 0;
    box-shadow: none;
}

/* avoid gaps between output lines */
div.nboutput div[class*=highlight] pre {
    line-height: normal;
}

/* input/output containers */
div.nbinput,
div.nboutput {
    display: -webkit-flex;
    display: flex;
    align-items: flex-start;
    margin: 0;
    width: 100%;
}
@media (max-width: 540px) {
    div.nbinput,
    div.nboutput {
        flex-direction: column;
    }
}

/* input container */
div.nbinput {
    padding-top: 5px;
}

/* last container */
div.nblast {
    padding-bottom: 5px;
}

/* input prompt */
div.nbinput div.prompt pre {
    color: #303F9F;
}

/* output prompt */
div.nboutput div.prompt pre {
    color: #D84315;
}

/* all prompts */
div.nbinput div.prompt,
div.nboutput div.prompt {
    min-width: 8ex;
    padding-top: 0.4em;
    padding-right: 0.4em;
    text-align: right;
    flex: 0;
}
@media (max-width: 540px) {
    div.nbinput div.prompt,
    div.nboutput div.prompt {
        text-align: left;
        padding: 0.4em;
    }
    div.nboutput div.prompt.empty {
        padding: 0;
    }
}

/* disable scrollbars on prompts */
div.nbinput div.prompt pre,
div.nboutput div.prompt pre {
    overflow: hidden;
}

/* input/output area */
div.nbinput div.input_area,
div.nboutput div.output_area {
    padding: 0.4em;
    -webkit-flex: 1;
    flex: 1;
    overflow: auto;
}
@media (max-width: 540px) {
    div.nbinput div.input_area,
    div.nboutput div.output_area {
        width: 100%;
    }
}

/* input area */
div.nbinput div.input_area {
    border: 1px solid #cfcfcf;
    border-radius: 2px;
    background: #f7f7f7;
}

/* override MathJax center alignment in output cells */
div.nboutput div[class*=MathJax] {
    text-align: left !important;
}

/* override sphinx.ext.pngmath center alignment in output cells */
div.nboutput div.math p {
    text-align: left;
}

/* standard error */
div.nboutput div.output_area.stderr {
    background: #fdd;
}

/* ANSI colors */
.ansi-black-fg { color: #3E424D; }
.ansi-black-bg { background-color: #3E424D; }
.ansi-black-intense-fg { color: #282C36; }
.ansi-black-intense-bg { background-color: #282C36; }
.ansi-red-fg { color: #E75C58; }
.ansi-red-bg { background-color: #E75C58; }
.ansi-red-intense-fg { color: #B22B31; }
.ansi-red-intense-bg { background-color: #B22B31; }
.ansi-green-fg { color: #00A250; }
.ansi-green-bg { background-color: #00A250; }
.ansi-green-intense-fg { color: #007427; }
.ansi-green-intense-bg { background-color: #007427; }
.ansi-yellow-fg { color: #DDB62B; }
.ansi-yellow-bg { background-color: #DDB62B; }
.ansi-yellow-intense-fg { color: #B27D12; }
.ansi-yellow-intense-bg { background-color: #B27D12; }
.ansi-blue-fg { color: #208FFB; }
.ansi-blue-bg { background-color: #208FFB; }
.ansi-blue-intense-fg { color: #0065CA; }
.ansi-blue-intense-bg { background-color: #0065CA; }
.ansi-magenta-fg { color: #D160C4; }
.ansi-magenta-bg { background-color: #D160C4; }
.ansi-magenta-intense-fg { color: #A03196; }
.ansi-magenta-intense-bg { background-color: #A03196; }
.ansi-cyan-fg { color: #60C6C8; }
.ansi-cyan-bg { background-color: #60C6C8; }
.ansi-cyan-intense-fg { color: #258F8F; }
.ansi-cyan-intense-bg { background-color: #258F8F; }
.ansi-white-fg { color: #C5C1B4; }
.ansi-white-bg { background-color: #C5C1B4; }
.ansi-white-intense-fg { color: #A1A6B2; }
.ansi-white-intense-bg { background-color: #A1A6B2; }

.ansi-default-inverse-fg { color: #FFFFFF; }
.ansi-default-inverse-bg { background-color: #000000; }

.ansi-bold { font-weight: bold; }
.ansi-underline { text-decoration: underline; }

/* CSS overrides for sphinx_rtd_theme */

/* 24px margin */
.nbinput.nblast,
.nboutput.nblast {
    margin-bottom: 19px;  /* padding has already 5px */
}

/* ... except between code cells! */
.nblast + .nbinput {
    margin-top: -19px;
}

.admonition > p:before {
    margin-right: 4px;  /* make room for the exclamation icon */
}
</style>
<div class="section" id="Analyze-2018-Pilot-runs-in-Parallel">
<h1>Analyze 2018 Pilot runs in Parallel<a class="headerlink" href="#Analyze-2018-Pilot-runs-in-Parallel" title="Permalink to this headline">¶</a></h1>
<div class="section" id="Shell-Variables">
<h2>Shell Variables<a class="headerlink" href="#Shell-Variables" title="Permalink to this headline">¶</a></h2>
<p>Assign the variables in this notebook.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre>
<span></span>In [1]:
</pre></div>
</div>
<div class="input_area highlight-bash notranslate"><div class="highlight"><pre>
<span></span>ls /data/hts2018_pilot/
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
ls: /data/hts2018_pilot/: No such file or directory
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>

</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre>
<span></span>In [2]:
</pre></div>
</div>
<div class="input_area highlight-bash notranslate"><div class="highlight"><pre>
<span></span><span class="nb">set</span> -u
<span class="nb">export</span> <span class="nv">CUROUT</span><span class="o">=</span><span class="nv">$HOME</span>/work/scratch/2018_stresstest
<span class="nb">export</span> <span class="nv">DATA_BASE</span><span class="o">=</span><span class="s2">&quot;/data/hts2018_pilot/&quot;</span>
<span class="nb">export</span> <span class="nv">RAW_FASTQS</span><span class="o">=</span><span class="s2">&quot;</span><span class="nv">$DATA_BASE</span><span class="s2">/Granek_4837_180427A5&quot;</span>
<span class="c1"># CURDATA_1=$CURDATA</span>
<span class="c1"># export INFO=$HOME/work/myinfo</span>
<span class="nb">export</span> <span class="nv">TRIMMED</span><span class="o">=</span><span class="nv">$CUROUT</span>/trimmed_fastqs
<span class="nb">export</span> <span class="nv">GENOME_DIR</span><span class="o">=</span><span class="nv">$CUROUT</span>/genome
<span class="nb">export</span> <span class="nv">STAR_OUT</span><span class="o">=</span><span class="nv">$CUROUT</span>/star_out_stress
<span class="nb">export</span> <span class="nv">ADAPTERS</span><span class="o">=</span><span class="nv">$DATA_BASE</span>/neb_E7600_adapters_withrc.fasta

<span class="c1"># ftp://ftp.ncbi.nlm.nih.gov/genomes/refseq/fungi/Cryptococcus_neoformans/all_assembly_versions/GCF_000149245.1_CNA3</span>
<span class="nb">export</span> <span class="nv">ACCESSION</span><span class="o">=</span><span class="s2">&quot;GCF_000149245.1_CNA3&quot;</span>
<span class="nb">export</span> <span class="nv">PREFIX</span><span class="o">=</span><span class="si">${</span><span class="nv">ACCESSION</span><span class="si">}</span>_genomic
<span class="nb">export</span> <span class="nv">GFF</span><span class="o">=</span><span class="si">${</span><span class="nv">PREFIX</span><span class="si">}</span>.gff
<span class="nb">export</span> <span class="nv">GTF</span><span class="o">=</span><span class="si">${</span><span class="nv">PREFIX</span><span class="si">}</span>.gtf
<span class="nb">export</span> <span class="nv">FNA</span><span class="o">=</span><span class="si">${</span><span class="nv">PREFIX</span><span class="si">}</span>.fna

<span class="nb">export</span> <span class="nv">FA_URL</span><span class="o">=</span><span class="s2">&quot;ftp://ftp.ensemblgenomes.org/pub/release-39/fungi/fasta/fungi_basidiomycota1_collection/cryptococcus_neoformans_var_grubii_h99/dna/Cryptococcus_neoformans_var_grubii_h99.CNA3.dna.toplevel.fa.gz&quot;</span>
<span class="nb">export</span> <span class="nv">GTF_URL</span><span class="o">=</span><span class="s2">&quot;ftp://ftp.ensemblgenomes.org/pub/release-39/fungi/gtf/fungi_basidiomycota1_collection/cryptococcus_neoformans_var_grubii_h99/Cryptococcus_neoformans_var_grubii_h99.CNA3.39.gtf.gz&quot;</span>
<span class="nb">export</span> <span class="nv">GFF_URL</span><span class="o">=</span><span class="s2">&quot;ftp://ftp.ensemblgenomes.org/pub/release-39/fungi/gff3/fungi_basidiomycota1_collection/cryptococcus_neoformans_var_grubii_h99/Cryptococcus_neoformans_var_grubii_h99.CNA3.39.gff3.gz&quot;</span>
<span class="nb">export</span> <span class="nv">GFF</span><span class="o">=</span><span class="k">$(</span>basename <span class="si">${</span><span class="nv">GFF_URL</span><span class="p">%.gz</span><span class="si">}</span><span class="k">)</span>
<span class="nb">export</span> <span class="nv">GTF</span><span class="o">=</span><span class="k">$(</span>basename <span class="si">${</span><span class="nv">GTF_URL</span><span class="p">%.gz</span><span class="si">}</span><span class="k">)</span>
<span class="nb">export</span> <span class="nv">FA</span><span class="o">=</span><span class="k">$(</span>basename <span class="si">${</span><span class="nv">FA_URL</span><span class="p">%.gz</span><span class="si">}</span><span class="k">)</span>
<span class="nb">export</span> <span class="nv">THREADS</span><span class="o">=</span><span class="m">14</span>
<span class="nb">export</span> <span class="nv">MAX_JOBS</span><span class="o">=</span><span class="m">1</span>

<span class="nb">echo</span> <span class="nv">$GFF</span> <span class="nv">$GTF</span> <span class="nv">$FA</span>

<span class="c1"># Making New Directories</span>
rm -rf <span class="nv">$CUROUT</span>
mkdir -p <span class="nv">$TRIMMED</span> <span class="nv">$GENOME_DIR</span> <span class="nv">$STAR_OUT</span>

<span class="c1"># ls $CURDATA</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Cryptococcus_neoformans_var_grubii_h99.CNA3.39.gff3 Cryptococcus_neoformans_var_grubii_h99.CNA3.39.gtf Cryptococcus_neoformans_var_grubii_h99.CNA3.dna.toplevel.fa
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre>
<span></span>In [3]:
</pre></div>
</div>
<div class="input_area highlight-bash notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># # ftp://ftp.ncbi.nlm.nih.gov/genomes/refseq/fungi/Cryptococcus_neoformans/all_assembly_versions/GCF_000149245.1_CNA3</span>
<span class="c1"># ftp://ftp.ensemblgenomes.org/pub/release-39/fungi/fasta/fungi_basidiomycota1_collection/cryptococcus_neoformans_var_grubii_h99/dna/Cryptococcus_neoformans_var_grubii_h99.CNA3.dna.toplevel.fa.gz</span>
<span class="c1"># ftp://ftp.ensemblgenomes.org/pub/release-39/fungi/gff3/fungi_basidiomycota1_collection/cryptococcus_neoformans_var_grubii_h99/Cryptococcus_neoformans_var_grubii_h99.CNA3.39.gff3.gz</span>

<span class="k">for</span> CUR in <span class="nv">$FA_URL</span> <span class="nv">$GTF_URL</span> <span class="p">;</span> <span class="k">do</span>
    wget --directory-prefix <span class="si">${</span><span class="nv">GENOME_DIR</span><span class="si">}</span> <span class="si">${</span><span class="nv">CUR</span><span class="si">}</span>
<span class="k">done</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
--2018-07-06 10:27:44--  ftp://ftp.ensemblgenomes.org/pub/release-39/fungi/fasta/fungi_basidiomycota1_collection/cryptococcus_neoformans_var_grubii_h99/dna/Cryptococcus_neoformans_var_grubii_h99.CNA3.dna.toplevel.fa.gz
           =&gt; ‘/Users/cliburn/work/scratch/2018_stresstest/genome/Cryptococcus_neoformans_var_grubii_h99.CNA3.dna.toplevel.fa.gz’
Resolving ftp.ensemblgenomes.org... 193.62.197.94
Connecting to ftp.ensemblgenomes.org|193.62.197.94|:21... connected.
Logging in as anonymous ... Logged in!
==&gt; SYST ... done.    ==&gt; PWD ... done.
==&gt; TYPE I ... done.  ==&gt; CWD (1) /pub/release-39/fungi/fasta/fungi_basidiomycota1_collection/cryptococcus_neoformans_var_grubii_h99/dna ... done.
==&gt; SIZE Cryptococcus_neoformans_var_grubii_h99.CNA3.dna.toplevel.fa.gz ... 5922212
==&gt; PASV ... done.    ==&gt; RETR Cryptococcus_neoformans_var_grubii_h99.CNA3.dna.toplevel.fa.gz ... done.
Length: 5922212 (5.6M) (unauthoritative)

Cryptococcus_neofor 100%[===================&gt;]   5.65M  1.38MB/s    in 4.3s

2018-07-06 10:27:50 (1.31 MB/s) - ‘/Users/cliburn/work/scratch/2018_stresstest/genome/Cryptococcus_neoformans_var_grubii_h99.CNA3.dna.toplevel.fa.gz’ saved [5922212]

--2018-07-06 10:27:50--  ftp://ftp.ensemblgenomes.org/pub/release-39/fungi/gtf/fungi_basidiomycota1_collection/cryptococcus_neoformans_var_grubii_h99/Cryptococcus_neoformans_var_grubii_h99.CNA3.39.gtf.gz
           =&gt; ‘/Users/cliburn/work/scratch/2018_stresstest/genome/Cryptococcus_neoformans_var_grubii_h99.CNA3.39.gtf.gz’
Resolving ftp.ensemblgenomes.org... 193.62.197.94
Connecting to ftp.ensemblgenomes.org|193.62.197.94|:21... connected.
Logging in as anonymous ... Logged in!
==&gt; SYST ... done.    ==&gt; PWD ... done.
==&gt; TYPE I ... done.  ==&gt; CWD (1) /pub/release-39/fungi/gtf/fungi_basidiomycota1_collection/cryptococcus_neoformans_var_grubii_h99 ... done.
==&gt; SIZE Cryptococcus_neoformans_var_grubii_h99.CNA3.39.gtf.gz ... 1796344
==&gt; PASV ... done.    ==&gt; RETR Cryptococcus_neoformans_var_grubii_h99.CNA3.39.gtf.gz ... done.
Length: 1796344 (1.7M) (unauthoritative)

Cryptococcus_neofor 100%[===================&gt;]   1.71M   337KB/s    in 6.0s

2018-07-06 10:27:57 (294 KB/s) - ‘/Users/cliburn/work/scratch/2018_stresstest/genome/Cryptococcus_neoformans_var_grubii_h99.CNA3.39.gtf.gz’ saved [1796344]

</pre></div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre>
<span></span>In [4]:
</pre></div>
</div>
<div class="input_area highlight-bash notranslate"><div class="highlight"><pre>
<span></span>gunzip --force <span class="si">${</span><span class="nv">GENOME_DIR</span><span class="si">}</span>/<span class="si">${</span><span class="nv">GTF</span><span class="si">}</span>.gz
gunzip --force <span class="si">${</span><span class="nv">GENOME_DIR</span><span class="si">}</span>/<span class="si">${</span><span class="nv">FA</span><span class="si">}</span>.gz
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre>
<span></span>In [5]:
</pre></div>
</div>
<div class="input_area highlight-bash notranslate"><div class="highlight"><pre>
<span></span>STAR <span class="se">\</span>
    --runThreadN <span class="nv">$THREADS</span> <span class="se">\</span>
    --runMode genomeGenerate <span class="se">\</span>
    --genomeDir <span class="nv">$GENOME_DIR</span> <span class="se">\</span>
    --genomeFastaFiles <span class="si">${</span><span class="nv">GENOME_DIR</span><span class="si">}</span>/<span class="si">${</span><span class="nv">FA</span><span class="si">}</span> <span class="se">\</span>
    --sjdbGTFfile <span class="si">${</span><span class="nv">GENOME_DIR</span><span class="si">}</span>/<span class="si">${</span><span class="nv">GTF</span><span class="si">}</span> <span class="se">\</span>
    --sjdbGTFfeatureExon exon <span class="se">\</span>
    --sjdbGTFtagExonParentTranscript transcript_id <span class="se">\</span>
    --sjdbGTFtagExonParentGene gene_id <span class="se">\</span>
    --genomeSAindexNbases <span class="m">6</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
STAR: Bad Option: --runThreadN.
Usage:  STAR cmd [options] [-find] file1 ... filen [find expression]

Use     STAR -help
and     STAR -xhelp
to get a list of valid cmds and options.

Use     STAR H=help
to get a list of valid archive header formats.

Use     STAR diffopts=help
to get a list of valid diff options.
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>

</pre></div></div>
</div>
<div class="section" id="Running-the-full-pipeline-on-all-the-samples">
<h3>Running the full pipeline on all the samples<a class="headerlink" href="#Running-the-full-pipeline-on-all-the-samples" title="Permalink to this headline">¶</a></h3>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre>
<span></span>In [6]:
</pre></div>
</div>
<div class="input_area highlight-bash notranslate"><div class="highlight"><pre>
<span></span>/usr/bin/time -v --format<span class="o">=</span><span class="s1">&#39;(%Xtext+%Ddata %Mmax)&#39;</span> ls
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
/usr/bin/time: illegal option -- v
usage: time [-lp] command.
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>

</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre>
<span></span>In [7]:
</pre></div>
</div>
<div class="input_area highlight-bash notranslate"><div class="highlight"><pre>
<span></span>trim_func<span class="o">()</span> <span class="o">{</span>
    <span class="c1"># echo $FASTQ</span>
    <span class="nv">FASTQ</span><span class="o">=</span><span class="nv">$1</span>
    <span class="nv">FASTQ</span><span class="o">=</span><span class="si">${</span><span class="nv">FASTQ</span><span class="p">##*/</span><span class="si">}</span> <span class="c1"># strip directory from file path</span>
    <span class="c1"># echo $FASTQ</span>
    <span class="nv">SAMPLE</span><span class="o">=</span><span class="s2">&quot;</span><span class="si">${</span><span class="nv">FASTQ</span><span class="p">%_L001_R1_001.fastq.gz</span><span class="si">}</span><span class="s2">&quot;</span> <span class="c1"># strip .fq.gz file extension</span>
    <span class="nb">echo</span> <span class="nv">$SAMPLE</span>
    <span class="c1"># exit 1</span>

<span class="c1">#     # gunzip all 4 lanes for each sample into</span>
<span class="c1">#     # run fastq-mcf on both miseq runs (e.g. concatenate on the fly) using a named pipe</span>
<span class="c1">#     # see https://github.com/ExpressionAnalysis/ea-utils/blob/wiki/FastqMcf.md</span>

<span class="c1">#     # make a pipe</span>
<span class="c1">#     CUR_PIPE=`mktemp --dry-run`_${SAMPLE}_pipe.fq</span>
<span class="c1">#     mkfifo $CUR_PIPE</span>
<span class="c1">#     # gunzip each sample&#39;s reads from all 4 lanes into the pipe</span>
<span class="c1">#     # This generates a concatenated stream of all 4 lanes</span>
<span class="c1">#     # $RAW_FASTQS/1_MA_*_L001_R1_001.fastq.gz</span>
<span class="c1">#     ls $RAW_FASTQS/${SAMPLE}_L00[1-4]_R1_001.fastq.gz</span>
<span class="c1">#     gunzip -c $RAW_FASTQS/${SAMPLE}_L00[1-4]_R1_001.fastq.gz &gt; $CUR_PIPE &amp;</span>

<span class="c1">#     # Use the pipe as input to fastq-mcf</span>
<span class="c1">#     fastq-mcf \</span>
<span class="c1">#         $ADAPTERS \</span>
<span class="c1">#         $CUR_PIPE \</span>
<span class="c1">#         -o $TRIMMED/${SAMPLE}.trim.fastq.gz \</span>
<span class="c1">#         -q 20 -x 0.5</span>

<span class="c1">#     rm -f $CUR_PIPE</span>

    /usr/bin/time -v --format<span class="o">=</span><span class="s1">&#39;(%Xtext+%Ddata %Mmax)&#39;</span> STAR <span class="se">\</span>
    --runMode alignReads <span class="se">\</span>
    --runThreadN <span class="nv">$THREADS</span> <span class="se">\</span>
    --readFilesCommand zcat <span class="se">\</span>
    --genomeDir <span class="nv">$GENOME_DIR</span> <span class="se">\</span>
    --outSAMtype BAM SortedByCoordinate <span class="se">\</span>
    --quantMode GeneCounts <span class="se">\</span>
    --genomeLoad LoadAndKeep <span class="se">\</span>
    --twopassMode None <span class="se">\</span>
    --limitBAMsortRAM <span class="m">1700000000</span> <span class="se">\</span>
    --outFileNamePrefix <span class="si">${</span><span class="nv">STAR_OUT</span><span class="si">}</span>/<span class="si">${</span><span class="nv">SAMPLE</span><span class="si">}</span> <span class="se">\</span>
    --outFilterScoreMinOverLread <span class="m">0</span> --outFilterMatchNminOverLread <span class="m">0</span> <span class="se">\</span>
    --outFilterMatchNmin <span class="m">0</span> --outFilterMismatchNmax <span class="m">2</span> <span class="se">\</span>
    --readFilesIn <span class="nv">$HOME</span>/work/scratch/2018_output/HTS_2018_pilot/trimmed_fastqs/<span class="si">${</span><span class="nv">SAMPLE</span><span class="si">}</span>.trim.fastq.gz
    <span class="c1"># -----------------------------------------------------</span>

<span class="o">}</span>
<span class="nb">export</span> -f trim_func


STAR --genomeDir <span class="nv">$GENOME_DIR</span> <span class="se">\</span>
    --genomeLoad LoadAndExit

<span class="c1"># parallel --jobs 1 pipeline_func ::: $RAW_FASTQS/1_MA_*_L001_R1_001.fastq.gz</span>
<span class="c1"># parallel --jobs $MAX_JOBS trim_func ::: $RAW_FASTQS/35_MA_*_L001_R1_001.fastq.gz</span>
<span class="c1"># parallel --jobs $MAX_JOBS trim_func ::: $RAW_FASTQS/*_L001_R1_001.fastq.gz</span>
parallel --jobs <span class="nv">$MAX_JOBS</span> trim_func ::: <span class="nv">$RAW_FASTQS</span>/3_TOT_J_S34_L001_R1_001.fastq.gz

STAR --genomeDir <span class="nv">$GENOME_DIR</span> <span class="se">\</span>
    --genomeLoad Remove
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
STAR: Bad Option: --genomeDir.
Usage:  STAR cmd [options] [-find] file1 ... filen [find expression]

Use     STAR -help
and     STAR -xhelp
to get a list of valid cmds and options.

Use     STAR H=help
to get a list of valid archive header formats.

Use     STAR diffopts=help
to get a list of valid diff options.
Academic tradition requires you to cite works you base your article on.
When using programs that use GNU Parallel to process data for publication
please cite:

  O. Tange (2011): GNU Parallel - The Command-Line Power Tool,
  ;login: The USENIX Magazine, February 2011:42-47.

This helps funding further development; AND IT WON&#39;T COST YOU A CENT.
If you pay 10000 EUR you should feel free to use GNU Parallel without citing.

To silence this citation notice: run &#39;parallel --citation&#39;.

3_TOT_J_S34
/usr/bin/time: illegal option -- v
usage: time [-lp] command.
STAR: Bad Option: --genomeDir.
Usage:  STAR cmd [options] [-find] file1 ... filen [find expression]

Use     STAR -help
and     STAR -xhelp
to get a list of valid cmds and options.

Use     STAR H=help
to get a list of valid archive header formats.

Use     STAR diffopts=help
to get a list of valid diff options.
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>

</pre></div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre>
<span></span>In [8]:
</pre></div>
</div>
<div class="input_area highlight-bash notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># ls $CUROUT/*</span>
<span class="c1"># head $CUROUT/star_out_stress/35_MA_P_S39Log.final.out</span>
<span class="c1"># rm -r $CUROUT</span>
</pre></div>
</div>
</div>
</div>
</div>
</div>


           </div>
           
          </div>
          <footer>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2018, Cliburn Chan.

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'../../',
            VERSION:'1.0',
            LANGUAGE:'None',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true,
            SOURCELINK_SUFFIX: '.txt'
        };
    </script>
      <script type="text/javascript" src="../../_static/jquery.js"></script>
      <script type="text/javascript" src="../../_static/underscore.js"></script>
      <script type="text/javascript" src="../../_static/doctools.js"></script>
      <script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>

  

  
  
    <script type="text/javascript" src="../../_static/js/theme.js"></script>
  

  <script type="text/javascript">
      jQuery(function () {
          
          SphinxRtdTheme.Navigation.enableSticky();
          
      });
  </script> 

</body>
</html>