.. HTS2018 documentation master file, created by
   sphinx-quickstart on Fri Jul  6 10:00:18 2018.
   You can adapt this file completely to your liking, but it should at least
   contain the root `toctree` directive.

HTS2018 resources
===================================

Source materials can bd cloned from the `HTS2018 GitLab Repository <https://gitlab.oit.duke.edu/HTS2018/HTS2018-notebooks.git>`_

.. toctree::
   :maxdepth: 2
   :caption: Slides

   slides.rst

.. toctree::
  :maxdepth: 2
  :caption: Unix Shell

  janice/1_IntroToCompBoot.ipynb
  janice/2_Intro_to_Unix_and_bash.ipynb
  cliburn/Bash_tutorial_prep.ipynb
  cliburn/The_Unix_Shell_01___File_and_Directory_Management.ipynb
  cliburn/The_Unix_Shell_02___Working_with_Text.ipynb
  cliburn/The_Unix_Shell_03___Finding_Stuff.ipynb
  cliburn/The_Unix_Shell_04___Regular_Expresssions.ipynb
  cliburn/The_Unix_Shell_05___Shell_Scripts.ipynb
  cliburn/The_Unix_Shell___Exercises_Solutions.ipynb
  cliburn/Bash_in_Jupyter.ipynb
  cliburn/Bash_Exercise_1_Solutions.ipynb
  cliburn/Bash_Exercise_2_Solutoins.ipynb
  # cliburn/Process-RNA-seq-counts.ipynb

.. toctree::
  :maxdepth: 2
  :caption: Bioinformatics

  bioinformatics/2018_bioinf_toc.ipynb
  bioinformatics/prep_nb/anon_metadata.ipynb
  bioinformatics/prep_nb/check_kouros_rpackages.ipynb
  bioinformatics/prep_nb/investigate_rrna_proportion.ipynb
  bioinformatics/prep_nb/star_2018_pilot.ipynb  bioinformatics/fastq_intro.ipynb
  bioinformatics/fastq_trimming.ipynb
  bioinformatics/fastqc.ipynb
  bioinformatics/genome_prep.ipynb
  bioinformatics/mapping.ipynb
  bioinformatics/quality_scores.ipynb
  bioinformatics/bash_functions.ipynb
  bioinformatics/bioinf_hands_on_1.ipynb
  bioinformatics/bioinf_hands_on_2.ipynb
  bioinformatics/bioinf_hands_on_3.ipynb
  bioinformatics/counting.ipynb
  bioinformatics/data_provenance.ipynb
  bioinformatics/demultiplex.ipynb
  bioinformatics/fastq_intro.ipynb
  bioinformatics/fastq_trimming.ipynb
  bioinformatics/fastqc.ipynb
  bioinformatics/genome_prep.ipynb
  bioinformatics/glob_loop.ipynb
  bioinformatics/igv_shorter_introns.ipynb
  bioinformatics/igv_visualization.ipynb
  bioinformatics/loop_pipeline.ipynb
  bioinformatics/making_a_pipeline.ipynb
  bioinformatics/making_generic_commands.ipynb
  bioinformatics/mapping.ipynb
  bioinformatics/multifastq_loop_pipeline.ipynb
  bioinformatics/paired_data_pipeline.ipynb
  bioinformatics/prep_for_full_run.ipynb
  bioinformatics/quality_scores.ipynb

.. toctree::
  :maxdepth: 2
  :caption: Analysis of Pilot Data

  pilot/01_Read_Scores.ipynb
  pilot/02_Plot_Counts.ipynb
  pilot/03_Compare_Gene_Expression.ipynb

.. toctree::
  :maxdepth: 2
  :caption: R Graphics and Data Manipulation

  janice/3_IntroductionToR.ipynb 
  cliburn/R_Graphics_Overview.ipynb
  cliburn/R_Graphics_Base.ipynb
  cliburn/R_Graphics_ggplot2.ipynb
  cliburn/R_Graphics_Exercise_Solutions.ipynb
  cliburn/R_tidyverse_1.ipynb
  cliburn/R_tidyverse_2.ipynb
  cliburn/R_tidyverse_3.ipynb
  cliburn/cliburn/R_tidyverse_Exercise_Solutions.ipynb

.. toctree::
  :maxdepth: 2
  :caption: R Statistical Analysis

  cliburn/Statistical_Inference_01.ipynb
  cliburn/Statistical_Inference_02_Solutions.ipynb
  cliburn/Unsupervised_And_Supervised_Learning.ipynb
  cliburn/Unsupervised_Learning.ipynb
  cliburn/Supervised_Learning Solutions.ipynb

.. toctree::
  :maxdepth: 2
  :caption: Bioconductor

  janice/4_Rlibraries_and_bioconductor.ipynb

.. toctree::
  :maxdepth: 2
  :caption: Biomedical Informatics

  biomedical_informatics/06_process_pathway_info.ipynb
  biomedical_informatics/07_Pathway_Analysis_and_Exploration.ipynb


Indices and tables
==================

* :ref:`genindex`
* :ref:`modindex`
* :ref:`search`
