{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Bioinformatics Table of Contents\n",
    "\n",
    "<!---\n",
    "### Lecture Slides\n",
    "* [Bioinformatics Pipeline Slides](bioinformatics_overview.pdf)\n",
    "\n",
    "#### TODO\n",
    "* Add intro text to each notebooks with overview of what it will cover\n",
    "* add in md5sum discussion (in QC?)\n",
    "\n",
    "\n",
    "data_provenance.ipynb\n",
    "quality_scores.ipynb\n",
    "-->\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "## Basic Bioinformatics Pipeline\n",
    "<!---\n",
    "1. [Demultiplex Raw FASTQs](filled_nb/demultiplex.ipynb) ([clean notebook](clean_nb/demultiplex_clean.ipynb))\n",
    "-->\n",
    "1. [Introduction to FASTQs](fastq_intro.ipynb)\n",
    "1. [Quality Control](fastqc.ipynb)\n",
    "1. [Trimming and Filtering](fastq_trimming.ipynb)\n",
    "1. [Download and Index Genome](genome_prep.ipynb)\n",
    "1. [Mapping Reads to a Reference Genome](mapping.ipynb)\n",
    "1. [Counting Reads](counting.ipynb)\n",
    "\n",
    "## Extending the Bioinformatics Pipeline\n",
    "1. [Working with Paired Reads](paired_data_pipeline.ipynb)\n",
    "1. [Making Generic Commands](making_generic_commands.ipynb)\n",
    "1. [Making a Pipeline](making_a_pipeline.ipynb)\n",
    "1. [Pipeline in a Loop](loop_pipeline.ipynb)\n",
    "1. [Running Multiple FASTQs](multifastq_loop_pipeline.ipynb)\n",
    "1. [Looping with Globs](glob_loop.ipynb)\n",
    "\n",
    "<!---\n",
    "1. [Pipeline with Loops](filled_nb/loop_pipeline.ipynb) ([clean notebook](clean_nb/loop_pipeline_clean.ipynb))\n",
    "1. [Pipeline for Multiple Samples](filled_nb/full_pilot_data_pipeline.ipynb) ([clean notebook](clean_nb/full_pilot_data_pipeline_clean.ipynb))\n",
    "\n",
    "## Pipeline for All 2015 Data\n",
    "\n",
    "1. [Demultiplex All 2015 Data](filled_nb/full_2015_data_demux.ipynb)\n",
    "1. [Demuxed FASTQ to Counts for All 2015 Data](filled_nb/full_2015_data_pipeline.ipynb)\n",
    "\n",
    "-->\n",
    "\n",
    "## Hands On Exercies\n",
    "1. [Hands On Exercise #1](bioinf_hands_on_1.ipynb)\n",
    "\n",
    "## Appendix\n",
    "* [Quality Scores](quality_scores.ipynb)\n",
    "\n",
    "<!---\n",
    "\n",
    "* [IGV visualization](filled_nb/igv_visualization.ipynb)\n",
    "\n",
    "-->\n",
    "\n",
    "<!---\n",
    "* [Using Bash in a Python Notebook](filled_nb/bash_magic.ipynb)\n",
    "\n",
    "## Parallelize\n",
    "* https://unix.stackexchange.com/questions/103920/parallelize-a-bash-for-loop\n",
    "* https://stackoverflow.com/questions/38160/parallelize-bash-script-with-maximum-number-of-processes\n",
    "* https://stackoverflow.com/questions/38774355/how-to-parallelize-for-loop-in-bash-limiting-number-of-processes\n",
    "-->\n",
    "\n",
    "<!---\n",
    "\n",
    "## Bonus\n",
    "* [Does Filtering Matter? Mapping unfiltered paired reads](filled_nb/paired_pilot_data_pipeline_unfiltered.ipynb)\n",
    "\n",
    "# Below Here Needs Checking\n",
    "1. [putting_it_together](filled_nb/putting_it_together.ipynb) ([clean notebook](clean_nb/putting_it_together_clean.ipynb))\n",
    "1. [demux_exploration](filled_nb/demux_exploration.ipynb) ([clean notebook](clean_nb/demux_exploration_clean.ipynb))\n",
    "1. [annotation_prep](filled_nb/annotation_prep.ipynb) ([clean notebook](clean_nb/annotation_prep_clean.ipynb))\n",
    "1. [putting_it_together_all_data](filled_nb/putting_it_together_all_data.ipynb) ([clean notebook](clean_nb/putting_it_together_all_data_clean.ipynb))\n",
    "1. [working_with_your_data](filled_nb/working_with_your_data.ipynb) ([clean notebook](clean_nb/working_with_your_data_clean.ipynb))\n",
    "1. [annotation](filled_nb/annotation.ipynb) ([clean notebook](clean_nb/annotation_clean.ipynb))\n",
    "\n",
    "#### Bioinformatics Pipeline on Full Dataset\n",
    "\n",
    "#### Differential Expression Analysis\n",
    "1. [DESeq2-Notebook](filled_nb/DESeq2-Notebook.ipynb) ([clean notebook](clean_nb/DESeq2-Notebook_clean.ipynb))\n",
    "1. [DESeq2-Notebook-from-Matrix](filled_nb/DESeq2-Notebook-from-Matrix.ipynb) ([clean notebook](clean_nb/DESeq2-Notebook-from-Matrix_clean.ipynb))\n",
    "\n",
    "### Additional Information\n",
    "* [Update Notebooks](filled_nb/update_notebooks.ipynb)\n",
    "* [Resources](https://github.com/granek/microbiome_workshop#resources)\n",
    "\n",
    "-->"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
