

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Bash Exercise 2: Working with a GTF file (Solutions) &mdash; HTS2018 1.0 documentation</title>
  

  
  
  
  

  

  
  
    

  

  
    <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="R Graphics" href="R_Graphics_Overview.html" />
    <link rel="prev" title="Bash Exercise 2: Working with a GTF file" href="Bash_Exercise_2.html" /> 

  
  <script src="../_static/js/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search">
          

          
            <a href="../index.html" class="icon icon-home"> HTS2018
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">Slides</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../slides.html">Slide sets</a></li>
</ul>
<p class="caption"><span class="caption-text">Unix Shell</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../janice/1_IntroToCompBoot.html">Duke HTS Summer Course 2018</a></li>
<li class="toctree-l1"><a class="reference internal" href="../janice/2_Intro_to_Unix_and_bash.html">Introduction to Unix and bash</a></li>
<li class="toctree-l1"><a class="reference internal" href="Bash_in_Jupyter.html">Using <code class="docutils literal notranslate"><span class="pre">bash</span></code> in bioinformatics</a></li>
<li class="toctree-l1"><a class="reference internal" href="Bash_Exercise_1.html">Bash Exercise</a></li>
<li class="toctree-l1"><a class="reference internal" href="Bash_Exercise_1_Solutions.html">Bash Exercise (Solutions)</a></li>
<li class="toctree-l1"><a class="reference internal" href="Bash_Exercise_2.html">Bash Exercise 2: Working with a GTF file</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Bash Exercise 2: Working with a GTF file (Solutions)</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#Optional-challenging-exercises">Optional challenging exercises</a></li>
</ul>
</li>
</ul>
<p class="caption"><span class="caption-text">R Graphics and Data Manipulation</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="R_Graphics_Overview.html">R Graphics</a></li>
<li class="toctree-l1"><a class="reference internal" href="R_Graphics_Base.html">R Graphics</a></li>
<li class="toctree-l1"><a class="reference internal" href="R_Graphic_ ggplot2.html">R Graphics</a></li>
<li class="toctree-l1"><a class="reference internal" href="R_Graphics_Exercise.html">R Graphics Exercise</a></li>
<li class="toctree-l1"><a class="reference internal" href="R_Graphics_Exercise_Solutions.html">R Graphics Exercise (Solutions)</a></li>
<li class="toctree-l1"><a class="reference internal" href="R_tidyverse_1.html">Using <code class="docutils literal notranslate"><span class="pre">dplyr</span></code> for data manipulation</a></li>
<li class="toctree-l1"><a class="reference internal" href="R_tidyverse_2.html">Using <code class="docutils literal notranslate"><span class="pre">tidyr</span></code> to create tidy data sets</a></li>
<li class="toctree-l1"><a class="reference internal" href="R_tidyverse_3.html">Working with multiple files</a></li>
<li class="toctree-l1"><a class="reference internal" href="R_tidyyverse_Exercise.html">R <code class="docutils literal notranslate"><span class="pre">tidyverse</span></code> exercise</a></li>
<li class="toctree-l1"><a class="reference internal" href="R_tidyyverse_Exercise_Solutions.html">R <code class="docutils literal notranslate"><span class="pre">tidyverse</span></code> exercise (Solutions)</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">HTS2018</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../index.html">Docs</a> &raquo;</li>
        
      <li>Bash Exercise 2: Working with a GTF file (Solutions)</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="../_sources/cliburn/Bash_Exercise_2_Solutoins.ipynb.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  
<style>
/* CSS for nbsphinx extension */

/* remove conflicting styling from Sphinx themes */
div.nbinput,
div.nbinput div.prompt,
div.nbinput div.input_area,
div.nbinput div[class*=highlight],
div.nbinput div[class*=highlight] pre,
div.nboutput,
div.nbinput div.prompt,
div.nbinput div.output_area,
div.nboutput div[class*=highlight],
div.nboutput div[class*=highlight] pre {
    background: none;
    border: none;
    padding: 0 0;
    margin: 0;
    box-shadow: none;
}

/* avoid gaps between output lines */
div.nboutput div[class*=highlight] pre {
    line-height: normal;
}

/* input/output containers */
div.nbinput,
div.nboutput {
    display: -webkit-flex;
    display: flex;
    align-items: flex-start;
    margin: 0;
    width: 100%;
}
@media (max-width: 540px) {
    div.nbinput,
    div.nboutput {
        flex-direction: column;
    }
}

/* input container */
div.nbinput {
    padding-top: 5px;
}

/* last container */
div.nblast {
    padding-bottom: 5px;
}

/* input prompt */
div.nbinput div.prompt pre {
    color: #303F9F;
}

/* output prompt */
div.nboutput div.prompt pre {
    color: #D84315;
}

/* all prompts */
div.nbinput div.prompt,
div.nboutput div.prompt {
    min-width: 8ex;
    padding-top: 0.4em;
    padding-right: 0.4em;
    text-align: right;
    flex: 0;
}
@media (max-width: 540px) {
    div.nbinput div.prompt,
    div.nboutput div.prompt {
        text-align: left;
        padding: 0.4em;
    }
    div.nboutput div.prompt.empty {
        padding: 0;
    }
}

/* disable scrollbars on prompts */
div.nbinput div.prompt pre,
div.nboutput div.prompt pre {
    overflow: hidden;
}

/* input/output area */
div.nbinput div.input_area,
div.nboutput div.output_area {
    padding: 0.4em;
    -webkit-flex: 1;
    flex: 1;
    overflow: auto;
}
@media (max-width: 540px) {
    div.nbinput div.input_area,
    div.nboutput div.output_area {
        width: 100%;
    }
}

/* input area */
div.nbinput div.input_area {
    border: 1px solid #cfcfcf;
    border-radius: 2px;
    background: #f7f7f7;
}

/* override MathJax center alignment in output cells */
div.nboutput div[class*=MathJax] {
    text-align: left !important;
}

/* override sphinx.ext.pngmath center alignment in output cells */
div.nboutput div.math p {
    text-align: left;
}

/* standard error */
div.nboutput div.output_area.stderr {
    background: #fdd;
}

/* ANSI colors */
.ansi-black-fg { color: #3E424D; }
.ansi-black-bg { background-color: #3E424D; }
.ansi-black-intense-fg { color: #282C36; }
.ansi-black-intense-bg { background-color: #282C36; }
.ansi-red-fg { color: #E75C58; }
.ansi-red-bg { background-color: #E75C58; }
.ansi-red-intense-fg { color: #B22B31; }
.ansi-red-intense-bg { background-color: #B22B31; }
.ansi-green-fg { color: #00A250; }
.ansi-green-bg { background-color: #00A250; }
.ansi-green-intense-fg { color: #007427; }
.ansi-green-intense-bg { background-color: #007427; }
.ansi-yellow-fg { color: #DDB62B; }
.ansi-yellow-bg { background-color: #DDB62B; }
.ansi-yellow-intense-fg { color: #B27D12; }
.ansi-yellow-intense-bg { background-color: #B27D12; }
.ansi-blue-fg { color: #208FFB; }
.ansi-blue-bg { background-color: #208FFB; }
.ansi-blue-intense-fg { color: #0065CA; }
.ansi-blue-intense-bg { background-color: #0065CA; }
.ansi-magenta-fg { color: #D160C4; }
.ansi-magenta-bg { background-color: #D160C4; }
.ansi-magenta-intense-fg { color: #A03196; }
.ansi-magenta-intense-bg { background-color: #A03196; }
.ansi-cyan-fg { color: #60C6C8; }
.ansi-cyan-bg { background-color: #60C6C8; }
.ansi-cyan-intense-fg { color: #258F8F; }
.ansi-cyan-intense-bg { background-color: #258F8F; }
.ansi-white-fg { color: #C5C1B4; }
.ansi-white-bg { background-color: #C5C1B4; }
.ansi-white-intense-fg { color: #A1A6B2; }
.ansi-white-intense-bg { background-color: #A1A6B2; }

.ansi-default-inverse-fg { color: #FFFFFF; }
.ansi-default-inverse-bg { background-color: #000000; }

.ansi-bold { font-weight: bold; }
.ansi-underline { text-decoration: underline; }

/* CSS overrides for sphinx_rtd_theme */

/* 24px margin */
.nbinput.nblast,
.nboutput.nblast {
    margin-bottom: 19px;  /* padding has already 5px */
}

/* ... except between code cells! */
.nblast + .nbinput {
    margin-top: -19px;
}

.admonition > p:before {
    margin-right: 4px;  /* make room for the exclamation icon */
}
</style>
<div class="section" id="Bash-Exercise-2:-Working-with-a-GTF-file-(Solutions)">
<h1>Bash Exercise 2: Working with a GTF file (Solutions)<a class="headerlink" href="#Bash-Exercise-2:-Working-with-a-GTF-file-(Solutions)" title="Permalink to this headline">¶</a></h1>
<p>Hack to handle broekn pipes in Jupyter - IGNORE.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre>
<span></span>In [14]:
</pre></div>
</div>
<div class="input_area highlight-bash notranslate"><div class="highlight"><pre>
<span></span>cleanup <span class="o">()</span> <span class="o">{</span>
    :
<span class="o">}</span>

<span class="nb">trap</span> <span class="s2">&quot;cleanup&quot;</span> SIGPIPE
</pre></div>
</div>
</div>
<p><strong>1</strong>. Create a subfolder called <code class="docutils literal notranslate"><span class="pre">refs</span></code> if it does not exist. You will
work in the <code class="docutils literal notranslate"><span class="pre">refs</span></code> folder for the rest of this exercise - i.e. make
this your working directory.</p>
<p>Download the compressed GTF file for <em>Cryptococcus neormans</em> from</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">ftp</span><span class="p">:</span><span class="o">//</span><span class="n">ftp</span><span class="o">.</span><span class="n">ensemblgenomes</span><span class="o">.</span><span class="n">org</span><span class="o">/</span><span class="n">pub</span><span class="o">/</span><span class="n">release</span><span class="o">-</span><span class="mi">39</span><span class="o">/</span><span class="n">fungi</span><span class="o">/</span><span class="n">gtf</span><span class="o">/</span><span class="n">fungi_basidiomycota1_collection</span><span class="o">/</span><span class="n">cryptococcus_neoformans_var_grubii_h99</span><span class="o">/</span><span class="n">Cryptococcus_neoformans_var_grubii_h99</span><span class="o">.</span><span class="n">CNA3</span><span class="o">.</span><span class="mf">39.</span><span class="n">gtf</span><span class="o">.</span><span class="n">gz</span>
</pre></div>
</div>
<p>to the <code class="docutils literal notranslate"><span class="pre">refs</span></code> folder maitaining its origanl name
<code class="docutils literal notranslate"><span class="pre">Cryptococcus_neoformans_var_grubii_h99.CNA3.39.gtf.gz</span></code>.</p>
<p>Note: You can use <code class="docutils literal notranslate"><span class="pre">wget</span></code> or <code class="docutils literal notranslate"><span class="pre">curl</span></code> to download the file.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre>
<span></span>In [1]:
</pre></div>
</div>
<div class="input_area highlight-bash notranslate"><div class="highlight"><pre>
<span></span>mkdir -p refs
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre>
<span></span>In [2]:
</pre></div>
</div>
<div class="input_area highlight-bash notranslate"><div class="highlight"><pre>
<span></span><span class="nb">cd</span> refs
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre>
<span></span>In [5]:
</pre></div>
</div>
<div class="input_area highlight-bash notranslate"><div class="highlight"><pre>
<span></span>wget ftp://ftp.ensemblgenomes.org/pub/release-39/fungi/gtf/fungi_basidiomycota1_collection/cryptococcus_neoformans_var_grubii_h99/Cryptococcus_neoformans_var_grubii_h99.CNA3.39.gtf.gz
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
--2018-07-03 10:23:35--  ftp://ftp.ensemblgenomes.org/pub/release-39/fungi/gtf/fungi_basidiomycota1_collection/cryptococcus_neoformans_var_grubii_h99/Cryptococcus_neoformans_var_grubii_h99.CNA3.39.gtf.gz
           =&gt; ‘Cryptococcus_neoformans_var_grubii_h99.CNA3.39.gtf.gz’
Resolving ftp.ensemblgenomes.org (ftp.ensemblgenomes.org)... 193.62.197.94
Connecting to ftp.ensemblgenomes.org (ftp.ensemblgenomes.org)|193.62.197.94|:21... connected.
Logging in as anonymous ... Logged in!
==&gt; SYST ... done.    ==&gt; PWD ... done.
==&gt; TYPE I ... done.  ==&gt; CWD (1) /pub/release-39/fungi/gtf/fungi_basidiomycota1_collection/cryptococcus_neoformans_var_grubii_h99 ... done.
==&gt; SIZE Cryptococcus_neoformans_var_grubii_h99.CNA3.39.gtf.gz ... 1796344
==&gt; PASV ... done.    ==&gt; RETR Cryptococcus_neoformans_var_grubii_h99.CNA3.39.gtf.gz ... done.
Length: 1796344 (1.7M) (unauthoritative)

Cryptococcus_neofor 100%[===================&gt;]   1.71M  1.95MB/s    in 0.9s

2018-07-03 10:23:37 (1.95 MB/s) - ‘Cryptococcus_neoformans_var_grubii_h99.CNA3.39.gtf.gz’ saved [1796344]

</pre></div></div>
</div>
<p><strong>2</strong>. Unzip the downloaded file. Use <code class="docutils literal notranslate"><span class="pre">ls</span></code> to see the contents of the
<code class="docutils literal notranslate"><span class="pre">refs</span></code> directory.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre>
<span></span>In [6]:
</pre></div>
</div>
<div class="input_area highlight-bash notranslate"><div class="highlight"><pre>
<span></span>gunzip Cryptococcus_neoformans_var_grubii_h99.CNA3.39.gtf.gz
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre>
<span></span>In [3]:
</pre></div>
</div>
<div class="input_area highlight-bash notranslate"><div class="highlight"><pre>
<span></span>ls
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Cryptococcus_neoformans_var_grubii_h99.CNA3.39.gtf
</pre></div></div>
</div>
<p><strong>3</strong>. Create a variable GTF for the filename. For bonus points, assign
GTF to the output of <code class="docutils literal notranslate"><span class="pre">ls</span></code>. Inspect the first 10 lines of this file.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre>
<span></span>In [4]:
</pre></div>
</div>
<div class="input_area highlight-bash notranslate"><div class="highlight"><pre>
<span></span><span class="nv">GTF</span><span class="o">=</span><span class="k">$(</span>ls<span class="k">)</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre>
<span></span>In [5]:
</pre></div>
</div>
<div class="input_area highlight-bash notranslate"><div class="highlight"><pre>
<span></span><span class="nb">echo</span> <span class="nv">$GTF</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Cryptococcus_neoformans_var_grubii_h99.CNA3.39.gtf
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre>
<span></span>In [6]:
</pre></div>
</div>
<div class="input_area highlight-bash notranslate"><div class="highlight"><pre>
<span></span>head -10 <span class="nv">$GTF</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
#!genome-build CNA3
#!genome-version CNA3
#!genome-date 2015-11
#!genome-build-accession GCA_000149245.3
#!genebuild-last-updated 2015-11
1       ena     gene    100     5645    .       -       .       gene_id &#34;CNAG_04548&#34;; gene_source &#34;ena&#34;; gene_biotype &#34;protein_coding&#34;;
1       ena     transcript      100     5645    .       -       .       gene_id &#34;CNAG_04548&#34;; transcript_id &#34;AFR92135&#34;; gene_source &#34;ena&#34;; gene_biotype &#34;protein_coding&#34;; transcript_source &#34;ena&#34;; transcript_biotype &#34;protein_coding&#34;;
1       ena     exon    5494    5645    .       -       .       gene_id &#34;CNAG_04548&#34;; transcript_id &#34;AFR92135&#34;; exon_number &#34;1&#34;; gene_source &#34;ena&#34;; gene_biotype &#34;protein_coding&#34;; transcript_source &#34;ena&#34;; transcript_biotype &#34;protein_coding&#34;; exon_id &#34;AFR92135-1&#34;;
1       ena     CDS     5494    5645    .       -       0       gene_id &#34;CNAG_04548&#34;; transcript_id &#34;AFR92135&#34;; exon_number &#34;1&#34;; gene_source &#34;ena&#34;; gene_biotype &#34;protein_coding&#34;; transcript_source &#34;ena&#34;; transcript_biotype &#34;protein_coding&#34;; protein_id &#34;AFR92135&#34;; protein_version &#34;1&#34;;
1       ena     start_codon     5643    5645    .       -       0       gene_id &#34;CNAG_04548&#34;; transcript_id &#34;AFR92135&#34;; exon_number &#34;1&#34;; gene_source &#34;ena&#34;; gene_biotype &#34;protein_coding&#34;; transcript_source &#34;ena&#34;; transcript_biotype &#34;protein_coding&#34;;
</pre></div></div>
</div>
<p><strong>4</strong>. Show only comment lines (those starting with <code class="docutils literal notranslate"><span class="pre">#</span></code>)</p>
<p>Hint: Use <code class="docutils literal notranslate"><span class="pre">grep</span></code> and an apppropriate regular expression.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre>
<span></span>In [7]:
</pre></div>
</div>
<div class="input_area highlight-bash notranslate"><div class="highlight"><pre>
<span></span>grep <span class="s1">&#39;^#&#39;</span> <span class="nv">$GTF</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
#!genome-build CNA3
#!genome-version CNA3
#!genome-date 2015-11
#!genome-build-accession GCA_000149245.3
#!genebuild-last-updated 2015-11
</pre></div></div>
</div>
<p><strong>5</strong>. A properly formatted GTF file has 9 columns separated by tabs
<code class="docutils literal notranslate"><span class="pre">\t</span></code>.</p>
<p>See <a class="reference external" href="https://www.gencodegenes.org/gencodeformat.html">ref</a> for what
information is in each column. From the linked site, here is a summary</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="mi">1</span>   <span class="n">chromosome</span> <span class="n">name</span> <span class="nb">chr</span><span class="p">{</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">4</span><span class="p">,</span><span class="mi">5</span><span class="p">,</span><span class="mi">6</span><span class="p">,</span><span class="mi">7</span><span class="p">,</span><span class="mi">8</span><span class="p">,</span><span class="mi">9</span><span class="p">,</span><span class="mi">10</span><span class="p">,</span><span class="mi">11</span><span class="p">,</span><span class="mi">12</span><span class="p">,</span><span class="mi">13</span><span class="p">,</span><span class="mi">14</span><span class="p">,</span><span class="mi">15</span><span class="p">,</span><span class="mi">16</span><span class="p">,</span><span class="mi">17</span><span class="p">,</span><span class="mi">18</span><span class="p">,</span><span class="mi">19</span><span class="p">,</span><span class="mi">20</span><span class="p">,</span><span class="mi">21</span><span class="p">,</span><span class="mi">22</span><span class="p">,</span><span class="n">X</span><span class="p">,</span><span class="n">Y</span><span class="p">,</span><span class="n">M</span><span class="p">}</span>
<span class="mi">2</span>   <span class="n">annotation</span> <span class="n">source</span>   <span class="p">{</span><span class="n">ENSEMBL</span><span class="p">,</span><span class="n">HAVANA</span><span class="p">}</span>
<span class="mi">3</span>   <span class="n">feature</span><span class="o">-</span><span class="nb">type</span>    <span class="p">{</span><span class="n">gene</span><span class="p">,</span><span class="n">transcript</span><span class="p">,</span><span class="n">exon</span><span class="p">,</span><span class="n">CDS</span><span class="p">,</span><span class="n">UTR</span><span class="p">,</span><span class="n">start_codon</span><span class="p">,</span><span class="n">stop_codon</span><span class="p">,</span><span class="n">Selenocysteine</span><span class="p">}</span>
<span class="mi">4</span>   <span class="n">genomic</span> <span class="n">start</span> <span class="n">location</span>  <span class="n">integer</span><span class="o">-</span><span class="n">value</span> <span class="p">(</span><span class="mi">1</span><span class="o">-</span><span class="n">based</span><span class="p">)</span>
<span class="mi">5</span>   <span class="n">genomic</span> <span class="n">end</span> <span class="n">location</span>    <span class="n">integer</span><span class="o">-</span><span class="n">value</span>
<span class="mi">6</span>   <span class="n">score</span> <span class="p">(</span><span class="ow">not</span> <span class="n">used</span><span class="p">)</span>    <span class="o">.</span>
<span class="mi">7</span>   <span class="n">genomic</span> <span class="n">strand</span>  <span class="p">{</span><span class="o">+</span><span class="p">,</span><span class="o">-</span><span class="p">}</span>
<span class="mi">8</span>   <span class="n">genomic</span> <span class="n">phase</span> <span class="p">(</span><span class="k">for</span> <span class="n">CDS</span> <span class="n">features</span><span class="p">)</span>    <span class="p">{</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="o">.</span><span class="p">}</span>
<span class="mi">9</span>   <span class="n">additional</span> <span class="n">information</span> <span class="k">as</span> <span class="n">key</span><span class="o">-</span><span class="n">value</span> <span class="n">pairs</span> <span class="p">(</span><span class="nb">format</span><span class="p">:</span> <span class="n">key</span> <span class="s2">&quot;value&quot;</span><span class="p">;)</span>
</pre></div>
</div>
<p>Show the first 6 lines of columns 1,3,4, and 5.</p>
<p>Hint: You can use <code class="docutils literal notranslate"><span class="pre">grep</span> <span class="pre">-v</span></code> to exclude unwanted lines.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre>
<span></span>In [15]:
</pre></div>
</div>
<div class="input_area highlight-bash notranslate"><div class="highlight"><pre>
<span></span>cat <span class="nv">$GTF</span> <span class="p">|</span>
grep -v <span class="s1">&#39;^#&#39;</span> <span class="p">|</span>
cut -f <span class="m">1</span>,3-5  <span class="p">|</span>
head -6
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
1       gene    100     5645
1       transcript      100     5645
1       exon    5494    5645
1       CDS     5494    5645
1       start_codon     5643    5645
1       exon    5322    5422
</pre></div></div>
</div>
<p><strong>6</strong>. Count how many of each feature (column 3) there are in the GTF.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre>
<span></span>In [17]:
</pre></div>
</div>
<div class="input_area highlight-bash notranslate"><div class="highlight"><pre>
<span></span>cat <span class="nv">$GTF</span> <span class="p">|</span> grep -v <span class="s1">&#39;^#&#39;</span> <span class="p">|</span> cut -f3 <span class="p">|</span> sort <span class="p">|</span> uniq -c
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
  49063 CDS
  52036 exon
   6923 five_prime_utr
   8497 gene
   7860 start_codon
   3167 stop_codon
   7034 three_prime_utr
   9348 transcript
</pre></div></div>
</div>
<p><strong>7</strong>. Show all the unique crhomosomes of the fungus, in ascending
numeric order.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre>
<span></span>In [29]:
</pre></div>
</div>
<div class="input_area highlight-bash notranslate"><div class="highlight"><pre>
<span></span>cat <span class="nv">$GTF</span> <span class="p">|</span> grep -v <span class="s1">&#39;^#&#39;</span> <span class="p">|</span> cut -f1 <span class="p">|</span> sort -n <span class="p">|</span> uniq
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Mt
1
2
3
4
5
6
7
8
9
10
11
12
13
14
</pre></div></div>
</div>
<p><strong>8</strong>. Extract all unique <code class="docutils literal notranslate"><span class="pre">gene_id</span></code> values found in column 9 and save
to a new file called <code class="docutils literal notranslate"><span class="pre">geen_counts.txt</span></code>. You can assume that
<code class="docutils literal notranslate"><span class="pre">geen_id</span></code> is always the first field in column 9. Remove double quotes
aorund gene_ids.</p>
<p>The first few lines of the file (shownn using <code class="docutils literal notranslate"><span class="pre">head</span></code>) looks lik this</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">CNAG_00001</span>
<span class="n">CNAG_00002</span>
<span class="n">CNAG_00003</span>
<span class="n">CNAG_00004</span>
<span class="n">CNAG_00005</span>
<span class="n">CNAG_00006</span>
<span class="n">CNAG_00007</span>
<span class="n">CNAG_00008</span>
<span class="n">CNAG_00009</span>
<span class="n">CNAG_00010</span>
</pre></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre>
<span></span>In [47]:
</pre></div>
</div>
<div class="input_area highlight-bash notranslate"><div class="highlight"><pre>
<span></span>cat <span class="nv">$GTF</span> <span class="p">|</span>             <span class="c1"># read file to standard input</span>
grep -v <span class="s1">&#39;^#&#39;</span> <span class="p">|</span>         <span class="c1"># exclude header lines</span>
cut -f9 <span class="p">|</span>              <span class="c1"># extract column 9</span>
cut -f1 -d<span class="s1">&#39;;&#39;</span> <span class="p">|</span>        <span class="c1"># extract column 1 using semi-colon as delimiter</span>
cut -f2 -d<span class="s1">&#39; &#39;</span> <span class="p">|</span>        <span class="c1"># extract column 2 using space as delimiter</span>
tr -d <span class="s1">&#39;&quot;&#39;</span> <span class="p">|</span>            <span class="c1"># remove double quotes</span>
sort <span class="p">|</span>                 <span class="c1"># sort in lexicographic order</span>
uniq &gt; geen_counts.txt <span class="c1"># save unique entries to text file</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre>
<span></span>In [46]:
</pre></div>
</div>
<div class="input_area highlight-bash notranslate"><div class="highlight"><pre>
<span></span>head geen_counts.txt
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
CNAG_00001
CNAG_00002
CNAG_00003
CNAG_00004
CNAG_00005
CNAG_00006
CNAG_00007
CNAG_00008
CNAG_00009
CNAG_00010
</pre></div></div>
</div>
<div class="section" id="Optional-challenging-exercises">
<h2>Optional challenging exercises<a class="headerlink" href="#Optional-challenging-exercises" title="Permalink to this headline">¶</a></h2>
<p>Use <code class="docutils literal notranslate"><span class="pre">awk</span></code>.</p>
<p><strong>9</strong>. Are there any non-header lines with more or less than 9
TAB-separated lines? How many non-header lines with exactly 9 columns
are ther? (Warning: This is quite challegning and an optional exercise)</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre>
<span></span>In [23]:
</pre></div>
</div>
<div class="input_area highlight-bash notranslate"><div class="highlight"><pre>
<span></span>grep -v <span class="s1">&#39;^#&#39;</span> <span class="nv">$GTF</span> <span class="p">|</span>
awk -F <span class="s1">&#39;\t&#39;</span> <span class="s1">&#39;{print NF}&#39;</span> <span class="p">|</span>
sort <span class="p">|</span>
uniq -c
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
 143928 9
</pre></div></div>
</div>
<p><strong>10</strong>. How many gene features are there in chromosome 2?</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre>
<span></span>In [24]:
</pre></div>
</div>
<div class="input_area highlight-bash notranslate"><div class="highlight"><pre>
<span></span>grep -v <span class="s1">&#39;^#&#39;</span> <span class="nv">$GTF</span> <span class="p">|</span>
awk -F <span class="s1">&#39;\t&#39;</span> <span class="s1">&#39;$1==&quot;2&quot; &amp;&amp; $3==&quot;gene&quot;&#39;</span>  <span class="p">|</span>
wc -l
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
706
</pre></div></div>
</div>
</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="R_Graphics_Overview.html" class="btn btn-neutral float-right" title="R Graphics" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="Bash_Exercise_2.html" class="btn btn-neutral" title="Bash Exercise 2: Working with a GTF file" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2018, Cliburn Chan.

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'../',
            VERSION:'1.0',
            LANGUAGE:'None',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true,
            SOURCELINK_SUFFIX: '.txt'
        };
    </script>
      <script type="text/javascript" src="../_static/jquery.js"></script>
      <script type="text/javascript" src="../_static/underscore.js"></script>
      <script type="text/javascript" src="../_static/doctools.js"></script>
      <script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>

  

  
  
    <script type="text/javascript" src="../_static/js/theme.js"></script>
  

  <script type="text/javascript">
      jQuery(function () {
          
          SphinxRtdTheme.Navigation.enableSticky();
          
      });
  </script> 

</body>
</html>